src/psyclone/psyir/nodes/structure_reference.py

```diff
@@ -233,13 +232,79 @@ def get_signature_and_indices(self):
 
         '''
         # Get the signature of self:
-        my_sig, my_index = \
-            super(StructureReference, self).get_signature_and_indices()
+        my_sig, my_index = super().get_signature_and_indices()
         # Then the sub-signature of the member, and indices used:
         sub_sig, indices = self.children[0].get_signature_and_indices()
         # Combine signature and indices
         return (Signature(my_sig, sub_sig), my_index + indices)
 
+    @property
+    def datatype(self):
+        '''
+        Walks down the list of members making up this reference to determine
+        the type that it refers to.
+
+        In order to minimise code duplication, this method also supports
+        ArrayOfStructuresReference by simply allowing for the case where
+        the starting reference is to an Array.
+
+        :returns: the datatype of this reference.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        '''
+        dtype = self.symbol.datatype
+
+        if isinstance(dtype, ArrayType):
+            dtype = dtype.intrinsic
+
+        if isinstance(dtype, DataTypeSymbol):
+            dtype = dtype.datatype
+
+        if isinstance(dtype, (DeferredType, UnknownType)):
+            # We don't know the type of the symbol that defines the type
+            # of this structure.
+            return dtype
```

Comment:
Oh yes, that revealed some problems. However, we can't have an array of DeferredType (since that implies that we have yet to resolve its type and therefore we don't know that it's an array). We *can* have arrays of UnknownType although that throws up problems in other places and also demonstrates why we shouldn't have the variable name in the text stored for the type. I've extended the testing.

Reply:
I'm still not sure what type to return if we hit a Deferred or Unknown type at any point. I'm thinking that the best thing to do is to always return DeferredType as UnknownType requires the text of the declaration and we don't have that. I've changed L266 to do that, pending your next review :-)

Reply:
Again, my confidence at what we should do at this point is low, so I won't ask it to be changed now, but I don't see any problem in having arrays of DeferredType. It just expresses our current knowledge of the type. For more we need to import the module information.